### ISSUE TITLE
[NEW] Support LZ4 compression

### ISSUE THREAD

**Issue by hpatro** [madolson/placeholderkv#1223]
**The problem/use-case that the feature addresses**

Faster `BGSAVE` time and smaller RDB snapshot size

**Description of the feature**

Support LZ4 compression as replacement of LZF compression for large in-memory data payload and RDB data. This should result in faster snapshot time as well as smaller RDB files. I had done experimentation in the past with the following [lz4 library](https://github.com/lz4/lz4) and the [performance improvement](https://github.com/redis/redis/issues/6809#issuecomment-1428517904) was really good with dummy benchmark data.

#### Data generation:

```
src/redis-benchmark -t set -n 1000000 -q -d 20000 -r 1000000
```

#### Performance:

Compression Type | BGSAVE time (sec) | RDB size (Mb) | load time (sec)
-- | -- | -- | --
LZ4 | 1089 | 8612.88 | 18.918
LZF | 1251 | 12405.38 | 32.34

It can be further fine tuned (I've not explored all the modes provided in the library).

The challenge we need to consider is supporting LZF compressed RDB data to be read (forever?) for backwards compatibliity if we decide to move to a new compression library.

**Comment by zuiderkwast** [comment]
IMO: We can add it when we bump the RDB version. This time we may want to change the magic string too since Redis already has defined RDB 12, so Valkey and Redis will diverge.

Q: Shall we vendor lz4?

**Comment by hpatro** [comment]
> IMO: We can add it when we bump the RDB version. This time we may want to change the magic string too since Redis already has defined RDB 12, so Valkey and Redis will diverge.

Yeah, that's a must. Based on RDB version we decompress data accordingly.

> Q: Shall we vendor lz4?

I think we should avoid that. git submodule is an option unless we need to make any changes to it.

**Comment by zuiderkwast** [comment]
I'm skeptical to submodules. Offline builds get complicated. Source releases get complicated. So far, we've used either vendored dependencies or system installed ones (like OpenSSL).

I prefer that we vendor this one. We could copy only one or a few files, as we've done with crc64 and some other small libraries. According to https://github.com/lz4/lz4/tree/dev/lib#readme it seems to be only one or a few files, depending on what we need.

**Comment by hpatro** [comment]
> I'm skeptical to submodules. Offline builds get complicated. Source releases get complicated. So far, we've used either vendored dependencies or system installed ones (like OpenSSL).
> 
> 
> 
> I prefer that we vendor this one. We could copy only one or a few files, as we've done with crc64 and some other small libraries. According to https://github.com/lz4/lz4/tree/dev/lib#readme it seems to be only one or a few files, depending on what we need.

I did that while experimenting but it's more maintenance work. I felt Ping generally had a stance of not vendoring packages into Valkey. Thoughts @PingXie ?

**Comment by zuiderkwast** [comment]
I do agree we should de-vendor large deps like Lua and Jemalloc and not vendor such deps in the future. For these, we can require them to be installed in the system, just like we do with OpenSSL, rather than trying to provide them.

For single file deps, that approach doesn't make much sense IMO. Small libs are often copied under src/, not even under deps/. Examples: crc64, lzf, MT19937-64 and siphash. The maintenance burden is not a problem for these, as far as I've seen. Not having them vendored would be a hazzle though.

**Comment by hpatro** [comment]
> I do agree we should de-vendor large deps like Lua and Jemalloc and not vendor such deps in the future. For these, we can require them to be installed in the system, just like we do with OpenSSL, rather than trying to provide them.
> 
> For single file deps, that approach doesn't make much sense IMO. Small libs are often copied under src/, not even under deps/. Examples: crc64, lzf, MT19937-64 and siphash. The maintenance burden is not a problem for these, as far as I've seen. Not having them vendored would be a hazzle though.

Do we get notified if there is a vulnerability discovered/patched in these libs you mentioned above?

**Comment by zuiderkwast** [comment]
No, only if we watch them in some way. Do we get notified if we use them as a submodule?

**Comment by asafpamzn** [comment]
In the example above, the value size used is 20KB, a size where compression is effective. However, for users working with smaller values—such as those under 256 bytes—compression algorithms are unlikely to be as effective.

If we are to modify the RDB protocol, I suggest compressing the RDB data in fixed-size chunks, such as 32KB, rather than compressing each value individually. This approach should improve the compression ratio and potentially reduce load times.

**Comment by hpatro** [comment]
> In the example above, the value size used is 20KB, a size where compression is effective. However, for users working with smaller values—such as those under 256 bytes—compression algorithms are unlikely to be as effective.
> 
> 
> 
> If we are to modify the RDB protocol, I suggest compressing the RDB data in fixed-size chunks, such as 32KB, rather than compressing each value individually. This approach should improve the compression ratio and potentially reduce load times.

Definitely, for RDB compression/decompression, this would make sense. 

For large objects stored in memory, if we are to apply compression, we would need to stick to the existing approach.

**Comment by PingXie** [comment]
> I did that while experimenting but it's more maintenance work. I felt Ping generally had a stance of not vendoring packages into Valkey. Thoughts @PingXie ?

We should avoid vendoring in general and gradually de-vendor existing dependencies, unless there's a compelling reason to do otherwise (e.g., the Lua case, where even minor version bumps break compatibility). That said, even for cases like Lua, I wonder if it might be worth exploring better ways to isolate breaking changes so that we could eventually devendor it as well, though I don’t have specific details at the moment.

With the adoption of the CMake build system, we now have robust tools for managing dependencies. Additionally, I think we can/should leverage [GitHub Dependabot](https://docs.github.com/en/code-security/supply-chain-security/keeping-your-dependencies-updated-automatically) to track changes in dependencies proactively. Dependabot creates pull requests for new dependency versions or patches and provides details such as changelogs and release notes when available. While Dependabot itself does not test for compatibility, it integrates seamlessly with CI pipelines, allowing us to automate testing of dependency updates to detect breakages early and manage updates efficiently.

**Comment by sarthakaggarwal97** [comment]
I was able to run multiple benchmarking experiments with both LZ4 and LZF. Sharing some of the experience with the forum. 

<img width="832" alt="Image" src="https://github.com/user-attachments/assets/401701f1-617a-4a8b-9eac-c8b7f76ec3a4" />

Notes: 

1. Acceleration factor of LZ4 does not impact the compressed data size or latency if all the values are same in the key. The performance worsened for random data.
2. The more redundant is the value, the better LZ4 seems to perform.


I captured the profiles during BGSAVE (there was no other operation on the server during that time).
Command to run profiling: `Command to run profiling: sudo perf record -p $(pidof valkey-server) -g -- sleep 60`

### LZ4:

<img width="705" alt="Image" src="https://github.com/user-attachments/assets/9a9118c2-cb04-4ea0-b1e5-42baf70af956" />


### LZF:

<img width="793" alt="Image" src="https://github.com/user-attachments/assets/5c4beb9d-4d74-4de4-9c76-ce0ba76168d3" />

It is observed that LZ4 does take fewer CPU cycles to compress than LZF, but overall improvements in load time and time to complete BGSAVE are visible in compressible data.

**Comment by hpatro** [comment]
This was an early effort for introducing LZ4 compression to Valkey. We will introduce it via #3195.
